File: apps/client/src-tauri/mobile-platform/android/src/main/java/MobilePlatformPlugin.kt
package boo.gchat.app.mobileplatform

import android.app.Activity
import app.tauri.annotation.Command
import app.tauri.annotation.InvokeArg
import app.tauri.annotation.TauriPlugin
import app.tauri.plugin.Invoke
import app.tauri.plugin.JSObject
import app.tauri.plugin.Plugin
import org.json.JSONObject
import java.util.concurrent.Executors

@InvokeArg
class SlotArgs { var slot: String = "" }

@InvokeArg
class StoreArgs {
    var slot: String = ""
    var secret: String = ""
    var confirmed: Boolean = false
}

@TauriPlugin
class MobilePlatformPlugin(private val activity: Activity) : Plugin(activity) {
    private val vault = UnlockVault(activity)

    companion object {
        // A process-wide serial worker preserves store/delete order, including
        // activity recreation. Keystore and filesystem work never blocks UI.
        private val worker = Executors.newSingleThreadExecutor { task ->
            Thread(task, "gchat-unlock-vault").apply { isDaemon = true }
        }
    }

    // Tauri has already loaded the GChat native library before registering this plugin.
    private external fun nativeLifecycle(state: Int)

    override fun onResume() { nativeLifecycle(1) }
    override fun onStop() {
        if (!activity.isChangingConfigurations) nativeLifecycle(2)
    }

    @Command
    fun getSecret(invoke: Invoke) {
        val args = try { invoke.parseArgs(SlotArgs::class.java) } catch (_: Exception) {
            invoke.reject("Invalid unlock storage request", "INVALID_ARGUMENT"); return
        }
        worker.execute {
            try {
                val secret = vault.read(args.slot)
                invoke.resolve(JSObject().put("secret", secret ?: JSONObject.NULL))
            } catch (_: Exception) { invoke.reject("Unlock storage is unavailable", "VAULT_UNAVAILABLE") }
        }
    }

    @Command
    fun storeSecret(invoke: Invoke) {
        val args = try { invoke.parseArgs(StoreArgs::class.java) } catch (_: Exception) {
            invoke.reject("Invalid unlock storage request", "INVALID_ARGUMENT"); return
        }
        worker.execute {
            try {
                vault.store(args.slot, args.secret, args.confirmed)
                invoke.resolve()
            } catch (_: Exception) { invoke.reject("Unlock storage is unavailable", "VAULT_UNAVAILABLE") }
            finally { args.secret = "" }
        }
    }

    @Command
    fun deleteSecret(invoke: Invoke) {
        val args = try { invoke.parseArgs(SlotArgs::class.java) } catch (_: Exception) {
            invoke.reject("Invalid unlock storage request", "INVALID_ARGUMENT"); return
        }
        worker.execute {
            try { vault.delete(args.slot); invoke.resolve() }
            catch (_: Exception) { invoke.reject("Unlock storage is unavailable", "VAULT_UNAVAILABLE") }
        }
    }
}

File: apps/client/src-tauri/mobile-platform/ios/Sources/MobilePlatformPlugin.swift
import Foundation
import Tauri
import UIKit
import WebKit

@_silgen_name("gchat_mobile_platform_lifecycle")
private func nativeLifecycle(_ state: UInt8)

private struct SlotArgs: Decodable { let slot: String }
private struct StoreArgs: Decodable { let slot: String; let secret: String; let confirmed: Bool }
private struct SecretReply: Encodable { let secret: String? }

final class MobilePlatformPlugin: Plugin {
    private let vault = UnlockVault()
    private static let worker = DispatchQueue(label: "boo.gchat.app.unlock-vault")
    private var observers: [NSObjectProtocol] = []

    @objc public override func load(webview: WKWebView) {
        guard observers.isEmpty else { return }
        let center = NotificationCenter.default
        observers.append(center.addObserver(forName: UIApplication.didEnterBackgroundNotification,
                                            object: nil, queue: .main) { _ in nativeLifecycle(2) })
        observers.append(center.addObserver(forName: UIApplication.didBecomeActiveNotification,
                                            object: nil, queue: .main) { _ in nativeLifecycle(1) })
        // Inactive (e.g. a transient system dialog) is not background.
        switch UIApplication.shared.applicationState {
        case .active: nativeLifecycle(1)
        case .background: nativeLifecycle(2)
        default: break
        }
    }

    deinit { observers.forEach(NotificationCenter.default.removeObserver) }

    @objc public func getSecret(_ invoke: Invoke) {
        let args: SlotArgs
        do { args = try invoke.parseArgs(SlotArgs.self) }
        catch { invoke.reject("Invalid unlock storage request", code: "INVALID_ARGUMENT"); return }
        Self.worker.async { [self] in
            do { invoke.resolve(SecretReply(secret: try vault.read(slot: args.slot))) }
            catch { invoke.reject("Unlock storage is unavailable", code: "VAULT_UNAVAILABLE") }
        }
    }

    @objc public func storeSecret(_ invoke: Invoke) {
        let args: StoreArgs
        do { args = try invoke.parseArgs(StoreArgs.self) }
        catch { invoke.reject("Invalid unlock storage request", code: "INVALID_ARGUMENT"); return }
        Self.worker.async { [self] in
            do { try vault.store(slot: args.slot, secret: args.secret, confirmed: args.confirmed); invoke.resolve() }
            catch { invoke.reject("Unlock storage is unavailable", code: "VAULT_UNAVAILABLE") }
        }
    }

    @objc public func deleteSecret(_ invoke: Invoke) {
        let args: SlotArgs
        do { args = try invoke.parseArgs(SlotArgs.self) }
        catch { invoke.reject("Invalid unlock storage request", code: "INVALID_ARGUMENT"); return }
        Self.worker.async { [self] in
            do { try vault.delete(slot: args.slot); invoke.resolve() }
            catch { invoke.reject("Unlock storage is unavailable", code: "VAULT_UNAVAILABLE") }
        }
    }
}

@_cdecl("init_plugin_gchat_mobile_platform")
func initPlugin() -> Plugin { MobilePlatformPlugin() }
